bolt-delegations-cli/src/types.rs

```diff
@@ -0,0 +1,106 @@
+use alloy::signers::k256::sha2::{Digest, Sha256};
+use ethereum_consensus::crypto::{PublicKey as BlsPublicKey, Signature as BlsSignature};
+use serde::Serialize;
+
+#[derive(Debug, thiserror::Error)]
+pub enum KeystoreError {
+    #[error("failed to read keystore directory: {0}")]
+    ReadFromDirectory(#[from] std::io::Error),
+    #[error("Failed to read or decrypt keystore: {0:?}")]
+    Eth2Keystore(lighthouse_eth2_keystore::Error),
+    #[error("Failed to get public key from keypair: {0}")]
+    UnknownPublicKey(String),
+}
+
+/// Event types that can be emitted by the validator pubkey to
+/// signal some action on the Bolt protocol.
+#[derive(Debug, Clone, Copy)]
+#[repr(u8)]
+#[allow(dead_code)]
+enum SignedMessageAction {
+    /// Signal delegation of a validator pubkey to a delegatee pubkey.
+    Delegation,
+    /// Signal revocation of a previously delegated pubkey.
+    Revocation,
+}
+
+/// Transparent serialization of signed messages.
+/// This is used to serialize and deserialize signed messages
+///
+/// e.g. serde_json::to_string(&signed_message):
+/// ```
+/// {
+///    "message": {
+///       "action": 0,
+///       "validator_pubkey": "0x...",
+///       "delegatee_pubkey": "0x..."
+///    },
+///   "signature": "0x..."
+/// },
+/// ```
+#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum SignedMessage {
+    Delegation(SignedDelegation),
+    Revocation(SignedRevocation),
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
+pub struct SignedDelegation {
+    pub message: DelegationMessage,
+    pub signature: BlsSignature,
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
+pub struct DelegationMessage {
+    action: u8,
+    pub validator_pubkey: BlsPublicKey,
+    pub delegatee_pubkey: BlsPublicKey,
+}
+
+impl DelegationMessage {
+    /// Create a new delegation message.
+    pub fn new(validator_pubkey: BlsPublicKey, delegatee_pubkey: BlsPublicKey) -> Self {
+        Self { action: SignedMessageAction::Delegation as u8, validator_pubkey, delegatee_pubkey }
+    }
+
+    /// Compute the digest of the delegation message.
+    pub fn digest(&self) -> [u8; 32] {
+        let mut hasher = Sha256::new();
+        hasher.update([self.action]);
+        hasher.update(self.validator_pubkey.to_vec());
+        hasher.update(self.delegatee_pubkey.to_vec());
+
+        hasher.finalize().into()
+    }
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
+pub struct SignedRevocation {
+    pub message: RevocationMessage,
+    pub signature: BlsSignature,
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
+pub struct RevocationMessage {
+    action: u8,
+    pub validator_pubkey: BlsPublicKey,
+    pub delegatee_pubkey: BlsPublicKey,
+}
```

Comment:
Do we need this distinction also in the codebase?

Reply:
it's really only useful when creating the messages, and we have it there